  $ \\ \huge{solution \ of \ 3.1 } \\ \text{ let  purchase   cost  of  A-1   A-2  ...... A-10  be} \ x_1 \  x_2 \  x_3 ........x_10 \  \text{and   in   house   cost   of   alloy   be   y_1   y_2 ........y_10} \\ \ \large{objective  \  \ function \  \  is \  \ :-} \\ \text{min} \\  82x_1 + 102x_2 + 150x_3 + 105x_4 + 95x_5 + 84x_6 + 88x_7 + 125x_8 + 70x_9 + 64x_10 + 45y_1 + 48y_2 +55y_3 + 65y_4 + 54y_5 + 37.5y_6 + 35y_7 + 52.5y_8 + 47y_9 + 28y_10 \\ \large{costraint \ \ is \ \ :-} \\  0.20x_1 + 0.25x_2 + 0.1x_3 + 0.3x_4 + 0.125x_5 + 0.35x_6 + 0.475x_7 + 0.1x_8 + 0.2x_9 + 0.5x_10 + 0.20y_1 + 0.25y_2 + 0.1y_3 + 0.3y_4 + 0.125y_5 + 0.35y_6 + 0.475y_7 + 0.1y_8 + 0.2y_9 + 0.5y_10 = 116.87500000000001
  \\  0.75x_1 + 0.65x_2 + 0.7x_3 + 0.6x_4 + 0.85x_5 + 0.525x_6 + 0.425x_7 + 0.4x_8 + 0.35x_9 + 0.4x_10 + 0.75y_1 + 0.65y_2 + 0.7y_3 + 0.6y_4 + 0.85y_5 + 0.525y_6 + 0.425y_7 + 0.4y_8 + 0.35y_9 + 0.4y_10 = 244.37499999999997 \\ 0.05x_1 + 0.1x_2 + 0.2x_3 + 0.1x_4 + 0.025x_5 + 0.125x_6 + 0.1x_7 + 0.5x_8 + 0.45x_9 + 0.1x_10 + 0.05y_1 + 0.1y_2 + 0.2y_3 + 0.1y_4 + 0.025y_5 + 0.125y_6 + 0.1y_7 + 0.5y_8 + 0.45y_9 + 0.1y_10 = 63.75 \\ \  \text{ lower bound of decision variable is [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]} \\ \text{upper bound of decision variable is [∞,∞,∞,∞,∞,∞,∞,∞,∞,∞,15,16,9,17.5,12,18.5,2.5,0.5,9.5,8]}$

In [ ]:
!pip install -q pyomo

     |████████████████████████████████| 9.6 MB 5.5 MB/s 
     |████████████████████████████████| 49 kB 6.0 MB/s 


In [ ]:
from pyomo.environ import *

In [ ]:
import numpy as np

In [ ]:
# no of variable
P = 20

In [ ]:
#no of constraint
Q = 3

In [ ]:
#create model
model3 = ConcreteModel()

In [ ]:
# declare objective coefficient
obj_coef_c1 = np.array([82,102,150,105,95,84,88,125,70,64,45,48,55,65,54,37.5,35,52.5,47,28])

In [ ]:
#declare constraint coefficient4
constr_coef_A1 = np.array([[0.2,0.25,0.1,0.3,0.125,0.35,0.475,0.1,0.2,0.5,0.2,0.25,0.1,0.3,0.125,0.35,0.475,0.1,0.2,0.5],[0.75,0.65,0.7,0.6,0.85,0.525,0.425,0.4,0.35,0.4,0.75,0.65,0.7,0.6,0.85,0.525,0.425,0.4,0.35,0.4],[0.05,0.1,0.2,0.1,0.025,0.125,0.1,0.5,0.45,0.1,0.05,0.1,0.2,0.1,0.025,0.125,0.1,0.5,0.45,0.1]])

In [ ]:
col_indices1 = np.arange(P)

In [ ]:
row_indices1 = np.arange(Q)

In [ ]:
model3.A=Var(col_indices1)

In [ ]:
model3.constraints=ConstraintList()

In [ ]:
constr_rhs_b1= np.array([116.87500000000001,244.37499999999997,63.75])

In [ ]:
lower_bound1=[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
upper_bound1 = [np.inf,np.inf,np.inf,np.inf,np.inf,np.inf,np.inf,np.inf,np.inf,np.inf,15,16,9,17.5,12,18.5,2.5,0.5,9.5,8]

In [ ]:
for i in col_indices1:
  model3.A[i].setlb(lower_bound1[i])
  model3.A[i].setub(upper_bound1[i])

In [ ]:
for j in row_indices1:
  model3.constraints.add(sum(model3.A[i]*constr_coef_A1[j][i] for i in col_indices1) == constr_rhs_b1[j])

In [ ]:
model3.objective = Objective( expr = sum(model3.A[i]*obj_coef_c1[i] for i in col_indices1),sense=minimize)

In [ ]:
model3.pprint()

2 Set Declarations
    A_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   20 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19}
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {1, 2, 3}

1 Var Declarations
    A : Size=20, Index=A_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  None : False :  True :  Reals
          1 :     0 :  None :  None : False :  True :  Reals
          2 :     0 :  None :  None : False :  True :  Reals
          3 :     0 :  None :  None : False :  True :  Reals
          4 :     0 :  None :  None : False :  True :  Reals
          5 :     0 :  None :  None : False :  True :  Reals
          6 :     0 :  None :  None : False :  True :  Reals
          7 :     0 :  None :  None : False :  True :  Reals
          8 :     

In [ ]:
!apt-get install -q -yy glpk-utils

Reading package lists...
Building dependency tree...
Reading state information...
The following package was automatically installed and is no longer required:
  libnvidia-common-460
Use 'apt autoremove' to remove it.
The following additional packages will be installed:
  libamd2 libcolamd2 libglpk40 libsuitesparseconfig5
Suggested packages:
  libiodbc2-dev
The following NEW packages will be installed:
  glpk-utils libamd2 libcolamd2 libglpk40 libsuitesparseconfig5
0 upgraded, 5 newly installed, 0 to remove and 19 not upgraded.
Need to get 692 kB of archives.
After this operation, 1,664 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu bionic/main amd64 libsuitesparseconfig5 amd64 1:5.1.2-2 [9,044 B]
Get:2 http://archive.ubuntu.com/ubuntu bionic/main amd64 libamd2 amd64 1:5.1.2-2 [19.5 kB]
Get:3 http://archive.ubuntu.com/ubuntu bionic/main amd64 libcolamd2 amd64 1:5.1.2-2 [16.2 kB]
Get:4 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libglpk40 amd6

In [ ]:
SolverFactory('glpk',excutable='/usr/bin/glpsol').solve(model3).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 28747.1875
  Upper bound: 28747.1875
  Number of objectives: 1
  Number of constraints: 4
  Number of variables: 21
  Number of nonzeros: 61
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.011530876159667969
# ----------------------------------------------------------
#   Solution Information
# ------------------------

In [ ]:
print('\n objective value',model3.objective())
print('\n variables')
for i in col_indices1:
  print('A[',i,'] = ',model3.A[i].value)
model3.constraints.display()


 objective value 28747.187500000007

 variables
A[ 0 ] =  158.078125
A[ 1 ] =  0.0
A[ 2 ] =  0.0
A[ 3 ] =  0.0
A[ 4 ] =  0.0
A[ 5 ] =  0.0
A[ 6 ] =  0.0
A[ 7 ] =  0.0
A[ 8 ] =  74.046875
A[ 9 ] =  84.3750000000001
A[ 10 ] =  15.0
A[ 11 ] =  16.0
A[ 12 ] =  9.0
A[ 13 ] =  17.5
A[ 14 ] =  12.0
A[ 15 ] =  18.5
A[ 16 ] =  2.5
A[ 17 ] =  0.5
A[ 18 ] =  9.5
A[ 19 ] =  8.0
constraints : Size=3
    Key : Lower              : Body               : Upper
      1 : 116.87500000000001 : 116.87500000000006 : 116.87500000000001
      2 : 244.37499999999997 :            244.375 : 244.37499999999997
      3 :              63.75 :              63.75 :              63.75


$ \Huge{solution \ \ of \ \ 3.3} \\ \huge{optimal \ \ objective \ \ function \ \  value \ \  =  \ \ 28747.187500000007} \\ \huge{optimal \ \ values \ \ of \ \  decision \ \ variable} \\  x_1 = 158.078125 \\ x_2 = 0 \\ x_3 = 0 \\ x_4 = 0 \\ x_5 =0 \\ x_6 = 0 \\ x_7 = 0 \\ x_8 = 0 \\ x_9 = 74.046875 \\ x_10 = 84.3750000000001 \\ y_1 =15.0 \\ y_2 = 16.0 \\ y_3 = 9.0 \\ y_4 = 17.5 \\ y_5 = 12.0 \\ y_6 = 18.5 \\ y_7 = 2.5 \\ y_8 = 0.5 \\ y_9 = 9.5 \\ y_10 = 8.0 $

$ \huge{solution \ \ \ of \ \ 3.4} \ \ \\ \\ \text{ A-1 > A-10 > A-9 ranked high in term of quantity} \\ \ \text{A-2 A-3 A-4 are ranked low}$

$ \Huge{solution \ \ of \ \ 3.5} \ \\ \text{yes A-1 A-2 A-3 A-4 A-5 A-6 A-7 A-8  A-9 A-10 STOCK are completely utilized} \\ \text{since in optimal solution we used all stock available of these alloy}$

$ \huge{solution \ \ of \ \ 3.6 } \\ \\ \text{yes A-2 A-3 A-4 A-5 A-6 A-7 A-8 need not be aquired from market} \\ \text{since in optimal solution  required criteria fullfill by in house stock} $